File: GrokkingAlgorithms/[Chapter4]quicksort.py
# def list_slicing(list):
#     return list[3:]

def quicksort(list):
    if len(list) < 2: return list

    else:
        pivot = list.pop(0)
        higher = []
        lower = []
        for element in list:
            lower.append(element) if element <= pivot else higher.append(element)
        return quicksort(lower) + [pivot] + quicksort(higher)

print(quicksort([10, 5, 2, 3]))

File: GrokkingAlgorithms/[Chapter4]quickSort.js
const quicksort = (unsortedArray) => {
  if (unsortedArray.length < 2) return unsortedArray;
  
  const pivot = unsortedArray[0];
  const newArray = unsortedArray.slice(1);
  const lower = [];
  const higher = [];
  
  newArray.forEach((element) =>
    element < pivot
    ? lower.push(element)
    : higher.push(element)
  );
  
  return quicksort(lower).concat([pivot]).concat(quicksort(higher))
};

console.log(quicksort([999, 3, 5, 2, 1, 4, 10, 1235, 25, 36, 7, 6, 98, 135, 27]));
console.log(quicksort(['fulano', 'xablau', 'cicrano']));
